primitives/src/adapter.rs

```diff
@@ -1,45 +1,52 @@
+use crate::channel::ChannelError;
 use crate::channel_validator::ChannelValidator;
 use crate::{Channel, DomainError, ValidatorId};
 use futures::future::BoxFuture;
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 use std::convert::From;
-use std::error::Error;
 use std::fmt;
-use std::fmt::Debug;
 
-pub type AdapterResult<T> = Result<T, AdapterError>;
+pub type AdapterResult<T, AE> = Result<T, Error<AE>>;
 
-#[derive(Debug, Eq, PartialEq)]
-pub enum AdapterError {
+pub trait AdapterErrorKind: fmt::Debug + fmt::Display {}
+
+#[derive(Debug)]
+pub enum Error<AE: AdapterErrorKind> {
     Authentication(String),
-    EwtVerifyFailed(String),
     Authorization(String),
-    Configuration(String),
-    Signature(String),
-    InvalidChannel(String),
-    Failed(String),
+    InvalidChannel(ChannelError),
+    /// Adapter specific errors
+    Adapter(AE),
+    Domain(DomainError),
+    /// You need to `.unlock()` the wallet first
+    LockedWallet,
 }
 
-impl Error for AdapterError {}
+impl<AE: AdapterErrorKind> std::error::Error for Error<AE> {}
 
-impl fmt::Display for AdapterError {
+impl<AE: AdapterErrorKind> From<AE> for Error<AE> {
+    fn from(adapter_err: AE) -> Self {
+        Self::Adapter(adapter_err)
+    }
+}
+
+impl<AE: AdapterErrorKind> fmt::Display for Error<AE> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            AdapterError::Authentication(error) => write!(f, "Authentication error: {}", error),
-            AdapterError::EwtVerifyFailed(error) => write!(f, "Ewt verification error: {}", error),
-            AdapterError::Authorization(error) => write!(f, "Authorization error: {}", error),
-            AdapterError::Configuration(error) => write!(f, "Configuration error: {}", error),
-            AdapterError::Signature(error) => write!(f, "Signature error: {}", error),
-            AdapterError::InvalidChannel(error) => write!(f, "{}", error),
-            AdapterError::Failed(error) => write!(f, "error: {}", error),
+            Error::Authentication(error) => write!(f, "Authentication error: {}", error),
```

Comment:
Do we really need the `error` word here? But if included  it should be of format camelcase
```
AuthenticationError: 
```

Reply:
Yep, we could remove them and leave only relevant context, as I did with the new errors. The PR got a bit bigger though so I stopped